src/api/proxy/CacheProxy.ts

```diff
@@ -0,0 +1,86 @@
+import { addSeconds } from 'date-fns'
+
+interface CacheObj<O> {
+  obj: O
+  createdOn: Date
+}
+
+export class CacheProxy<T> {
+  protected api: T
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private cache: { [hash: string]: CacheObj<any> }
+
+  public constructor(api: T) {
+    this.api = api
+    this.cache = {}
+  }
+
+  /**
+   * Returns value from cache for given `hash`.
+   * Cache behaviour is given by `expiration` property:
+   *  0: return cache if found == cache forever
+   *  number: return cache if found and cache younger than now + `expiration` seconds
+   */
+  private get<R>(hash: string, expiration: number): R | undefined {
+    const cacheObj = this.cache[hash]
+
+    if (!cacheObj) {
+      // there's no cache
+      return
+    }
+
+    if (expiration > 0) {
+      const validUntil = addSeconds(new Date(), expiration)
+
+      if (cacheObj.createdOn > validUntil) {
+        // the cache is expired
+        return
+      }
+    }
+
+    // cache was found
+    console.debug(`cache hit for ${hash}`)
+    return cacheObj.obj
+  }
+
+  private store<O>(hash: string, obj: O): void {
+    this.cache[hash] = { obj, createdOn: new Date() }
+  }
+
+  protected async fetchWithCache<P, R, M = keyof T>(method: M, params: P, expiration?: number): Promise<R> {
```

Comment:
Is this a good idea to mark method as async? That would make some sync cached functions unnecessarily sync

Reply:
Did not think much about it now because the apis I want to cache are all async (contract calls).

But yeah, this did bother me, but can I use the same method for both sync and async versions?

Reply:
I think async is fine.
By definition, we want to cache stuff that takes time otherwise, and if it takes time should be async.

I'd leave it async until we find an issue, then we can think if we just create a `fetchWithCacheSync`  or whatever solution we came up with

Reply:
> but can I use the same method for both sync and async versions?

don't await, cache a promise and return that promise

Reply:
Ohhhh, cache the promise! Didn't know that would work. Trivial change, will try it out.

Reply:
Done :)